app/services/reporting/agency_and_sp_report.rb

```diff
@@ -48,30 +48,36 @@ def agency_and_sp_emailable_report
     end
 
     def active_agencies
-      @active_agencies ||= Agreements::PartnerAccountStatus.find_by(name: 'active').
-        partner_accounts.
-        includes(:agency).
-        where('became_partner <= ?', report_date).
-        map(&:agency).
-        uniq
+      @active_agencies ||= Agency.joins(:partner_accounts).
+        where(partner_accounts: {
+          partner_account_status: Agreements::PartnerAccountStatus.find_by(name: 'active'),
+          became_partner: ..report_date,
+        }).
+        distinct
     end
 
     def service_providers
-      @service_providers ||= Reports::BaseReport.transaction_with_timeout do
-        issuers = ServiceProviderIdentity.
-          where('created_at <= ?', report_date).
-          distinct.
-          pluck(:service_provider)
-        ServiceProvider.where(issuer: issuers).active.external
+      return @service_providers if defined?(@service_providers)
+      issuers = ServiceProvider.active.external.pluck(:issuer).filter_map do |issuer|
+        exists = Reports::BaseReport.transaction_with_timeout do
+          SpReturnLog.where(billable: true, issuer: issuer).where(
+            'returned_at::date <= ?', report_date
+          ).exists?
+        end
+
+        issuer if exists
       end
```

Comment:
1. would it make sense to do a `#select` instead of a `#filter_map`? seems like the result type is the same as the input?

```suggestion
      issuers = ServiceProvider.active.external.pluck(:issuer).select do |issuer|
        Reports::BaseReport.transaction_with_timeout do
          SpReturnLog.where(billable: true, issuer: issuer).where(
            'returned_at::date <= ?', report_date
          ).exists?
        end
      end
```

2. is there any way we can do this all in one query and still take advantage of the index?

Reply:
2. I did a little bit of testing, and it times out to query all of them and group by the issuer. It does use the index, but I suspect having to do an `ANY` query with hundreds of issuers makes it hard to use it well?

Reply:
After some further investigation, the existing query seems to be faster so I've reverted the related changes.